src/dmd/dcast.d

```diff
@@ -1485,23 +1485,38 @@ extern (C++) Expression castTo(Expression e, Scope* sc, Type t)
             if (AggregateDeclaration t1ad = isAggregate(t1b))
             {
                 AggregateDeclaration toad = isAggregate(tob);
-                if (t1ad != toad && t1ad.aliasthis)
+
+                if (t1ad != toad)
                 {
-                    if (t1b.ty == Tclass && tob.ty == Tclass)
+                    bool tryaliasthis = true;
+                    if (tob.ty == Tstruct && (cast(TypeStruct)t1b).sym == (cast(TypeStruct)tob).sym)
+                    {
+                        tryaliasthis = false;
+                    }
+                    else if (t1b.ty == Tclass && tob.ty == Tclass)
                     {
                         ClassDeclaration t1cd = t1b.isClassHandle();
                         ClassDeclaration tocd = tob.isClassHandle();
                         int offset;
                         if (tocd.isBaseOf(t1cd, &offset))
-                            goto Lok;
+                             goto Lok;
                     }
+                    if (tryaliasthis)
+                    {
+                        Expression[] results;
+                        iterateAliasThis(sc, e, &TypeAliasThisCtx(t).castTo, results);
```

Comment:
Here, introduction of a complex block to handle alias this resolving where it was once only a call to `resolveAliasThis()`.